### DISCUSSION TITLE
Send MIDI data on startup

### DISCUSSION THREAD

**Discussion by chrisF808** [tttapa/Control-Surface#512] [Q&A]
Hello, I am currently experimenting with the Control Surface library. The library is very clear and everything works wonderfully. I can move the virtual faders in the DAW and in the light control program with my physical controllers (faders) on the Arduino.

I would like the MIDI controller to poll its faders once when the program starts and send the MIDI data to the DAW so that the positions of the faders are identical in hardware and software. 

Is there a way to query the physical faders and send the MIDI data corresponding to the analog value once when the Arduino starts?

many greetings, Christian

**Comment by MarcosGogo** [comment]
I don't understand much, but I think there might be some workaround with this [MIDI-Output example](https://github.com/tttapa/Control-Surface/blob/master/examples/3.%20MIDI%20Interfaces/MIDI-Output/MIDI-Output.ino), taking out the LOOP information and changing the setup to your fader addressing and some [GetValue](https://tttapa.github.io/Control-Surface-doc/Doxygen/d5/d8e/classMIDIFilteredAnalog.html#a401ab4ab293a05ef4b9d33c8b55472ca) command for the potentiometer, or [pitchbend sender class](https://tttapa.github.io/Control-Surface-doc/Doxygen/de/d46/classPitchBendSender.html).

But I'm pretty new to programming and with this library. So I can be completely wrong.

**Comment by tttapa** [comment]
I've just added a [`forcedUpdate()`](https://tttapa.github.io/Control-Surface-doc/Doxygen/d5/d8e/classMIDIFilteredAnalog.html#a1bf4c3dda90b58bb483c1ddaae6be5e1) method to the potentiometer and absolute encoder classes in d08e4228cf244406a271ca2473b1c7123a021166. You can use it as follows:

```cpp
#include <Control_Surface.h>

USBDebugMIDI_Interface midi;

CCPotentiometer pots[] {
  {A0, 0},
  {A1, 1},
  {A2, 2},
  {A3, 3},
};

void setup() {
  Control_Surface.begin();
  while (!Serial);
  for (auto &pot : pots)
    pot.forcedUpdate();
}

void loop() {
  Control_Surface.loop();
}
```

**Reply by MarcosGogo** [reply]
> I've just added a [`forcedUpdate()`](https://tttapa.github.io/Control-Surface-doc/Doxygen/d5/d8e/classMIDIFilteredAnalog.html#a1bf4c3dda90b58bb483c1ddaae6be5e1) method to the potentiometer and absolute encoder classes in [d08e422](https://github.com/tttapa/Control-Surface/commit/d08e4228cf244406a271ca2473b1c7123a021166)

Nice. Will this make it possible to use bank with AbsoluteEncoder without position memory when switching banks?

**Reply by tttapa** [reply]
> Nice. Will this make it possible to use bank with AbsoluteEncoder without position memory when switching banks?

I'm not sure what you mean by that, but I'm afraid the answer is no. This change adds no new functionality, it just makes it easier to (re-)send MIDI messages from your main program. You'll see that the implementation of `forcedUpdate()` is just a single line: https://github.com/tttapa/Control-Surface/commit/d08e4228cf244406a271ca2473b1c7123a021166#diff-d0de2dfc74def088993b89e51a028b2b53afd44e60148518771238c1219b151fR64

**Reply by chrisF808** [reply]
> I've just added a [`forcedUpdate()`](https://tttapa.github.io/Control-Surface-doc/Doxygen/d5/d8e/classMIDIFilteredAnalog.html#a1bf4c3dda90b58bb483c1ddaae6be5e1) method to the potentiometer and absolute encoder classes in [d08e422](https://github.com/tttapa/Control-Surface/commit/d08e4228cf244406a271ca2473b1c7123a021166). You can use it as follows:
> 
> ```c++
> #include <Control_Surface.h>
> 
> USBDebugMIDI_Interface midi;
> 
> CCPotentiometer pots[] {
>   {A0, 0},
>   {A1, 1},
>   {A2, 2},
>   {A3, 3},
> };
> 
> void setup() {
>   Control_Surface.begin();
>   while (!Serial);
>   for (auto &pot : pots)
>     pot.forcedUpdate();
> }
> 
> void loop() {
>   Control_Surface.loop();
> }
> ```

This works very well, I delayed the call a bit more, now all virtual sliders jump to the position of the hardware sliders.
Thanks for your answer and your work!

**Reply by jglinis** [reply]
I'm trying to implement this same idea but over Hardware Serial MIDI.

Substituting 
```cpp
HardwareSerialMIDI_Interface midi {Serial1};

```
for

```cpp
USBDebugMIDI_Interface midi;

```

works as long as I have usb connected to my laptop and the serial monitor is open, but I plan to use this as a standalone 5 pin DIN MIDI controller. 

If I change

```cpp
void setup() {
  Control_Surface.begin();
  while (!Serial);
  for (auto &pot : pots)
    pot.forcedUpdate();
}

```

to correspond with my Hardware interface Serial1,

```cpp
  Control_Surface.begin();
   while (Serial1);
  for (auto &pot : pots)
    pot.forcedUpdate();
}

```
then the sketch will compile, but nothing works. 
Here's a stripped down version of my sketch, just trying to get this to work for one potentiometer:


```cpp
int led = LED_BUILTIN;

#include <Control_Surface.h>

HardwareSerialMIDI_Interface midi {Serial1};


CCPotentiometer pots[] {
  {A0, {36, Channel_3}},
};

void setup() {
  // put your setup code here, to run once:
  pinMode(led, OUTPUT);

  Control_Surface.begin();
   while (Serial1);
  for (auto &pot : pots)
    pot.forcedUpdate();
}

void loop() {
  digitalWrite(led, HIGH);

  Control_Surface.loop();
}
```

Any suggestions? Thanks!

**Reply by jglinis** [reply]
Ah. I just remove the while loop altogether. I could swear I tried this and it didn't work. Anyway, looks to be working now!
